command line analysis tools support

joelagnel opened the pull request:
Hi,
Just posting some early patches to get feedback on the command line analysis tools feature. Forgive me for not adding proper docstrings which I'll add when posting follow ups (hence tagged RFC).

The idea is you can  post-process traces or other data in a results directory, and get useful information from them without having to launch a browser or notebook environment. It fits very well with LISA's analysis architecture (libs/utils/analysis/)
Some data is simple enough that it shouldn't need plotting. Also incase of results that need to be made more visual, I am planning to add some code that autogenerates a .html file by running a dynamically generated notebook and automatically launch a browser window to show them) - for example for analysis tools that do need to show something more visual.

Thanks,
Joel

joelagnel commented:
CC @bjackman @derkling

bjackman commented on libs/utils/trace.py:
```diff
@@ -256,16 +256,17 @@ def __parseTrace(self, path, tasks, window, normalize_time, trace_format):
-        self._sanitize_SchedOverutilized()
-        self._sanitize_CpuFrequency()
+        # Santization not possible if platform missing
+        if not self.platform:
```
Is this meant to be be `if self.platform`?

derkling replied:
I completely agree with the commit message title, but this is like using an axe instead of a bistoury... even when `platform` data are not available we can still (and have to) do many of the trace sanitation updates.
I would better move this check inside each and every `_sanitize_XXX` method to carefully cut out what is not possible.

At the same time, this update requires also to cross check all the analysis and plotting APIs to do the same.
Indeed, right now we assume that we have all the columns that are added by the availability of platform data. This should be done carefully and, since it's a big job by itself, I would rather suggest to these updates on a PR by their own.

Can you remove this patch from this series for the time being while I work on a separate PR to properly remove the platform data dependency? I've just created #427 to track this cleanups.

joelagnel replied:
Ok I'll drop it from this PR

bjackman commented on libs/utils/analysis_tool.py:
```diff
@@ -0,0 +1,83 @@
+        self.trace = Trace(None, args.trace, [])
+
+	self.output = self.run_analysis()
+	self.run_output()
```
This might just be my own weird obsession but I don't like when `__init__` does things other than up data structures - I'd much prefer to have everything ultimately driven by a `analysis_main` function or something, which is called from inside an `if __name__ == "__main__":` - it seems a lot more flexible.

derkling replied:
This is ultimately defining an "execution model" for a generic analysis.

We should make it as much generic as possible, maybe something like this:

```python
class BatchAnalysis():
    def __init__():
        # command line parsing
    def parse():
        # Trace parsing and DF generation
    def analyze():
        # Trace post-processing and results generation
    def generate():
        # Output data generation
    def run():
        self.parse()
        self.analyze()
        self.generate()
```

So that a client class usually re_define `parse`, `analyze` and `generate` while the `run` method is the one called by default from the `__main__` of an executable script which is hosted in `tools/analysis`.

joelagnel replied:
@derkling just wanted some clarification:
*  in your example above, you mentioned 'command line parsing' in `__init__`. But then in the `run` method you have `self.parse` again. Going by @bjackman comment about not doing "actions" in `__init__`, I am wondering if that's what you meant.
* the "executable script" will also host the tool-specific `parse`, `analyze` and `generate` which are overridden in the specific tool. Hope we are on the same page about that?

derkling commented on libs/utils/analysis_tool.py:
```diff
@@ -0,0 +1,83 @@
+        print self.output
+
+    def __init__(self, parser=None, gui=False):
+	self.gui = gui
```
Mind TAB vs spaces... we use always and only spaces...

derkling commented on libs/utils/analysis_tool.py:
```diff
@@ -0,0 +1,83 @@
+        if not args.log_stdout:
+            LisaLogging.setup(level=logging.ERROR)
+        else:
+            LisaLogging.setup()
```
```python
level = logging.INFO if args.log_stdout else logging.ERROR
LisaLogging.setup(level=level)
```

derkling commented on libs/utils/analysis_tool.py:
```diff
@@ -0,0 +1,83 @@
+
+        # Create trace object
+        self._log.info('Creating trace object...')
+        self.trace = Trace(None, args.trace, [])
```
Perhaps this requires a TRAPpy extension to parse by default all the events, right?
If that's the case I would suggest to add a patch which makes `events` an optional parameter for the `Trace` ctor.

joelagnel replied:
Sure, I'll make it optional

derkling commented on libs/utils/analysis_tool.py:
```diff
@@ -0,0 +1,83 @@
+            raise NotImplementedError('error')
+        print self.output
+
+    def __init__(self, parser=None, gui=False):
```
Maybe better `s/parser/options_parser`?
Moreover, what's an example for the usage of the `gui` option?

joelagnel replied:
Will add proper docstrings, sorry for making the review hard. GUI is supposed to be for tools that want to show output in the browser instead of text. The idea is for certain tools to create a report automatically (using ipython notebook generation utilities) and automatically open it in the browser.

derkling commented on libs/utils/analysis_tool.py:
```diff
@@ -0,0 +1,83 @@
+    def run_analysis(self):
+	raise NotImplementedError('error')
+
+    def run_output(self):
```
What's the semantic of this method?

valschneider replied:
Since this class is intended to be subclassed, I think it would be nice to add a docstring to all of the "abstract" methods.

joelagnel replied:
I'll follow up with better docstrings, sorry. `run_output` is supposed to present output to the user. Every tool will have its own output. Maybe we can provide some generic templates later.

joelagnel replied:
fixed

derkling commented on tools/analysis/android_gfx:
```diff
@@ -0,0 +1,39 @@
+from analysis_tool import AnalysisTool
+
+
+class GfxAnalysisTool(AnalysisTool):
```
I would suggest to use the same name for both the class and the file. The name should also be "kind of" namespace based and details. Thus, I would say that `class AndroidGFX` should be better.

Moreover, I guess this is just an initial simple example and that you wanna add much more GFX related results to be generated by running this script, isn't it?

In that case I would suggest to feed-in some more example to better verify if the API works for even just slightly more complex use-cases.

joelagnel replied:
Yes this is just an initial example to flesh out the architecture. Will change class naming as you suggest 👍

valschneider commented on tools/analysis/android_gfx:
```diff
@@ -0,0 +1,39 @@
+from time import sleep
+from devlib.exception import TargetError
+from analysis_tool import AnalysisTool
+
```
Nitpicking here, there's one more line feed than is needed.

joelagnel replied:
fixed

joelagnel commented:
Hi @derkling @bjackman 
Please check this and let me know if it looks Ok. Please note:
* I dropped 'make platform optional' patch as @derkling wants to do that so that's needed for these patches to work. But I did add a 'make events optional' patch as @derkling requested.
* I kept the graphics example simple because I don't want to complicate it right now - as I said, this is a simple example.
* I dropped the gui stuff since I didn't want to complicate it and add something not yet in use.

derkling commented on libs/utils/analysis_tool.py:
```diff
@@ -0,0 +1,116 @@
+        self.trace = Trace(None, self.args.trace)
+
+        self.run_analysis()
+        self.run_output()
```
Why do we wanna enforce `run_output` and `run_analysis` to be two different functions it then we call them one after the other? I would say that from the standpoint of the "state machine" defined by the case class they are exactly the same state. Thus I would keep just one of the two methods... sub-classes can still split the implementation in two if they want and/or find it useful.

Moreover, I'm not convinced about the naming of these last three methods. I would prefer something more "state/action like"... what about `AnalysisTool::analyse` and `AnalysisTool::run`... where the user is required to put all of its logic and (optional) output production in the `analyse` method.

joelagnel replied:
I would still want `run_analysis` but I am Ok with dropping output if that's ok :-) deal ?

joelagnel replied:
So I think I still want `run_output` now that I think about it more. The reason is in the future if we do any special formatting of results than just print, then it becomes more manageable to do it in one place than modify every script. Second reason is it allows a tool to separate out analysis and reporting logic by just overriding a single function so it's cleaner. I actually don't think it adds much more complexity anyway.

Hope we can agree on something here or maybe @bjackman can make a suggestion and we take the majority vote ;-)

derkling replied:
Now every script produce its output in a single `self.output` variable, which is printed by the default (dummy) implementation of `run_output()`... if in the future you should need any "special formatting" it's quite likely you still need to update all scripts because the simple `self.output` is... too simple.

Do you have any special use-case already in mind? In that case you should provide an example, otherwise your arguments now are not convincing to me.

Regarding `AnalysisTool::run_analysis` vs `AnalysisTool::analyse` I cannot see why you are so much into to that name?
Is there any other reason behind just the name itself?
Otherwise I think we should use @bjackman's vote on this as well ;-)

joelagnel replied:
Ok. Let's drop output call back then, for now and let individual script do printing and formatting.
Regarding the name, it's a personal preference.. I guess to me it sounds more explicit that you're execute something when you say run_. I don't have a strong argument here and I'm ok with @bjackman vote.

bjackman replied:
As for naming I think both `run_analysis` and `analyse` are fine.. probably lean towards `run_analysis` if it gives output too, for the same reason as Joel i.e. `analyse` sort of sounds like it returns some representation of a conclusion whereas `run_analysis` sounds like it has a side effect (i.e. printing output).

derkling replied:
Then let's got for `run_analysis`... ;-)

joelagnel replied:
Cool ;-)

derkling commented on tools/analysis/android_gfx:
```diff
@@ -0,0 +1,39 @@
+
+    def run_analysis(self):
+        df = self.trace.data_frame.trace_event('tracing_mark_write')
+
```
You can skip this empty line... and if you go for the removal of the `run_output` method, here you just go straight for a print statement ;-)

derkling commented on libs/utils/trace.py:
```diff
@@ -73,7 +73,7 @@ class Trace(object):
     """
 
-    def __init__(self, platform, data_dir, events,
+    def __init__(self, platform, data_dir, events=None,
```
:heart_eyes: another step into the direction to make `Trace()` super simple to build, thanks! ;-)

I still have to provide a patch to make `platform` optional as well... will do soon.

joelagnel replied:
Cool, thanks

joelagnel replied:
Cool, thanks

bjackman commented on libs/utils/analysis_tool.py:
```diff
@@ -0,0 +1,116 @@
+class AnalysisTool(object):
+    """
+    A base class for LISA analysis command-line tools.
+    This class intended to be subclassed in order to create a custom
```
Also, empty line after the first line please - https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings (I know we probably don't follow this properly but it's nice to try :P)

joelagnel replied:
Cool. :-) happy to learn these conventions

bjackman commented on libs/utils/analysis_tool.py:
```diff
@@ -0,0 +1,116 @@
+                required=True)
+
+        parser.add_argument('--log-stdout', action='store_true',
+                default=False,
```
Don't think you need the `default=False` when `action='store_true'`

joelagnel replied:
Ok

bjackman commented on libs/utils/analysis_tool.py:
```diff
@@ -0,0 +1,116 @@
+
+        # General options that all tools accept
+        parser.add_argument('--trace', type=str,
+                default=None,
```
No need for `default` when `required=True`

joelagnel replied:
Ok got it, thanks

bjackman commented on libs/utils/analysis_tool.py:
```diff
@@ -0,0 +1,116 @@
+        The specific analysis tool should override this and perform
+        analysis and provide results in self.output
+        """
+        raise NotImplementedError('error')
```
Add a slightly more helpful message here?

joelagnel replied:
Sure

bjackman commented on libs/utils/analysis_tool.py:
```diff
@@ -0,0 +1,116 @@
+
+        # Create trace object
+        self._log.info('Creating trace object...')
+        self.trace = Trace(None, self.args.trace)
```
Github makes it look like this commit comes before the one that makes `events` optional. If that's true, please re-order them. [Github might be misleading me though :angry: :sob: :disappointed: :nauseated_face: ](https://help.github.com/articles/why-are-my-commits-in-the-wrong-order/)

joelagnel replied:
Ok

bjackman commented on tools/analysis/android_gfx:
```diff
@@ -0,0 +1,39 @@
+        self.output = {'frame_drops': self.get_frame_drops(df)}
+
+# Run the analysis and provide results
+AndroidGfx().run_main()
```
Should put this inside an `if __name__ == "__main__":`, that way it gets run when you execute this script from the cmdline but not when you import it from elsewhere.

joelagnel replied:
Since I am a python n00b relatively speaking... Does this convention also apply to when you don't expect the scripts to be imported?

derkling replied:
I would say it's more clean (or "pythonic")... even if not strictly required if it will never be included.
But at least it makes clear it is providing a directly executable functionality...

joelagnel replied:
Got it.will fix

joelagnel commented on libs/utils/analysis_tool.py:
```diff
@@ -0,0 +1,116 @@
+    analysis tool.
+
+    :param options_parser: A parser object passed by the specific tool
+                           to parse specific options if any (optional).
```
Ok

joelagnel commented on libs/utils/analysis_tool.py:
```diff
@@ -0,0 +1,116 @@
+
+        Analysis tools shouldn't output to standard output as their
+        output can be used by other tools. Configure the logger to
+        output only error messages
```
Ok will do

joelagnel commented on libs/utils/analysis_tool.py:
```diff
@@ -0,0 +1,116 @@
+        tool specific options are also parsed
+        """
+        parser = self.options_parser
+
```
Ok

joelagnel commented:
Got it. Will fix

joelagnel commented:
@derkling fixed all review comments, could you merge?

bjackman commented:
Just been taking another look at this and doing some thinking - I wonder if instead of just printing the results, the AnalysisTools should

bjackman commented:
Sorry, I cannot github

bjackman commented:
What I meant to say was: I wonder if these AnalysisTools should, rather than printing out analysis results directly, return them in a standardized format (maybe just a dict) that can then be formatted in a flexible way (e.g. pretty-print to the console for interactive use but perhaps also save JSON for use in CI)?

The example I'm thinking of is a command-line tool to expose what I've added here: https://github.com/ARM-software/lisa/pull/455. I'd like to return a few different statistics about underutilization, and it would be nice if I know they'll always be formatted in the same way as the other analysis tools.

I might be over-thinking this.

joelagnel commented:
Hi @bjackman
About your last comment about returning dict than printing, the main usecase I had in mind when I wrote this is PR is that we should print because these tools can be used by other programs to easily parse the machine-readable results. My thought was any analysis results that internal to LISA should be returned by libs/utils/analysis as a DF. The tools/analysis/ stuff is supposed to be just a front end. I'm going to restart this work and post another PR soon,

joelagnel commented:
Rebased and got rid of init as @bjackman suggested. I think this one has been beaten to death :) lets merge.

bjackman commented:
> we should print because these tools can be used by other programs to easily parse the machine-readable results.

Yeah, I agree, but I was thinking this would be easier to do if the actual printing is done in a common way. So instead of each analysis tool doing the print, it returns the data in some python representation (dict/DF/something else), then the base class prints it. So then if we wanted to change the format of printing (or have multiple formats, e.g. JSON/CSV) we wouldn't have to change each results processor.

One of the cool ways we could use this is to integrate with [workload-automation](https://github.com/ARM-software/workload-automation) so maybe LISA could be used as a "results processor" for that.

joelagnel commented:
Hi @bjackman I am ok with changing to making the 'printing' done by the Base class, that way in the future if we have to change it to something else, we only change 1 place. Let me know if this is what you want. I think this is what I was doing in my earlier approaches but I was suggested to do the printing in the specific tool itself.

Speaking in terms of working well with the broader LISA scheme of things, I believe the analysis classes in libs/utils/analysis/ should be the only ones to return DF/dict/whatever strictly, all the others should be the reporters/consumers/presenters. The exception is if the results from those classes need further processing. This is just my thought about delineating things more strictly

bjackman commented:
OK, that's interesting. Yeah that is what I was suggesting but maybe @derkling disagrees. Perhaps we need to have a cal to discuss this in a little more detail...
